pex/cli/commands/lockfile/updater.py

```diff
@@ -0,0 +1,205 @@
+            lock_configuration=lock_configuration,
+            original_requirements=original_requirements,
+            pip_configuration=pip_configuration,
+        )
+
+    original_requirements = attr.ib()  # type: Iterable[str]
+    update_constraints_by_project_name = attr.ib()  # type: Mapping[ProjectName, Requirement]
+    lock_configuration = attr.ib()  # type: LockConfiguration
+    pip_configuration = attr.ib()  # type: PipConfiguration
+
+    @contextmanager
+    def _calculate_update_constraints(self, locked_resolve):
+        # type: (LockedResolve) -> Iterator[Optional[Iterable[str]]]
+        if not self.update_constraints_by_project_name:
+            yield None
+            return
+
+        constraints = []
+        for locked_requirement in locked_resolve.locked_requirements:
+            pin = locked_requirement.pin
+            constraint = self.update_constraints_by_project_name.get(
+                pin.project_name, pin.as_requirement()
+            )
+            constraints.append(str(constraint))
+        if not constraints:
+            yield None
+            return
+
+        with named_temporary_file(prefix="lock_update.", suffix=".constraints.txt", mode="w") as fp:
+            fp.write(os.linesep.join(constraints))
+            fp.flush()
+            try:
+                yield [fp.name]
+            except ResultError as e:
+                logger.error(
+                    "The following lock update constraints could not be satisfied:\n"
+                    "{constraints}".format(constraints="\n".join(constraints))
+                )
+                raise e
+
+    def _log(self, message):
+        if self.dry_run:
+            print(message)
+        else:
+            logger.info(message)
+
+    def update_resolve(
+        self,
+        locked_resolve,  # type: LockedResolve
+        target_configuration,  # type: TargetConfiguration
+    ):
+        # type: (...) -> Union[LockUpdate, Error]
+
+        with self._calculate_update_constraints(locked_resolve) as constraints_files:
+            updated_lock_file = try_(
+                create(
+                    lock_configuration=self.lock_configuration,
+                    requirement_configuration=RequirementConfiguration(
+                        requirements=self.original_requirements,
+                        constraint_files=constraints_files,
+                    ),
+                    target_configuration=target_configuration,
+                    pip_configuration=self.pip_configuration,
+                )
+            )
+
+        assert 1 == len(updated_lock_file.locked_resolves)
+        updated_resolve = updated_lock_file.locked_resolves[0]
+
+        updated_requirements_by_project_name = OrderedDict(
+            (updated_requirement.pin.project_name, updated_requirement)
+            for updated_requirement in updated_resolve.locked_requirements
+        )  # type: OrderedDict[ProjectName, LockedRequirement]
```

Comment:
Will this handle deleting a `LockedRequirement` if it is no longer needed?

Reply:
Yes. That's covered by this test: https://github.com/pantsbuild/pex/pull/1478/files#diff-8c268962cf5d39934d82d9fe9fe14fc7032ef0df2f946c5077d4c40ca0578bdfR310